### PR TITLE
Feature/i_jepa_model

### PR DESCRIPTION
**Pull request by zassou65535**
## 概要

I-JEPAのEncoderとPredictor(引用元: [I-JEPAの本家リポジトリ](https://github.com/facebookresearch/ijepa) )、それらのテストの追加

## Submit前の確認項目

- [x] タイトルは一目でわかるようにし、説明文は PR を簡潔に説明するようにしましたか?
- [x] あなたの PR が、異なる変更を束ねたものではなく、ひとつのことだけを行うものであることを確認しましたか?
- [x] この PR で導入されたすべての変更点をリストアップしましたか?
- [x] PR を`make run`コマンドでローカルでテストしましたか?

## 補足

気になった点がいくつか
- Maskの型に関して
    - [I-JEPAの本家リポジトリ](https://github.com/facebookresearch/ijepa) では、CollatorやModel部分とかで各Maskは`List[torch.Tensor]`型(各`Tensor`は`[batch_size, n_patches(取り出したいpatchの数)]`)と扱われているけど、[マスクを適用する関数`apply_masks`の実装](https://github.com/facebookresearch/ijepa/blob/main/src/masks/utils.py)を見ると、Maskに含まれる各`Tensor`の`n_patches`は全部同じなことを前提にしているっぽい？(1つでも`n_patches`が違う`Tensor`が含まれていると最後の`return torch.cat(all_x, dim=0)`でconcatに失敗する)
        - もしかするとI-JEPAではMaskは全部`List[torch.Tensor]`じゃなくて`torch.Tensor`として扱ってしまっても良いのかも？(ただ、もしこれ関係をリファクタリングするとしたら別PRでやるべきな気がするから、とりあえずこのPRでは本家に倣って`List[torch.Tensor]`としておく)
- Test用パラメーターの共通化
    - このPRではTestとして`test_vision_transformer_encoder`関数と`test_vision_transformer_predictor`関数があるけど、`@pytest.mark.parametrize`で指定してるパラメーターに、2つで共通しているものが多いから共通化しようかと思ったけどいまいちいい実装が思い浮かばず迷ってる　そのせいで実装が冗長になってるかも
        - 何かいい実装案あったら教えて

### CODE REVIEW

**Review by Geson-anko — CHANGES_REQUESTED** [review]
実装ありがとう。だけどPRがデカすぎてレビューが大変。少なくとも前処理系、モデル定義で分離して別々にPRを送ろう。 ブランチ名を `feature/#<issue番号>/branch-name`にすると良い。

テストは一つに統合するよりも処理が透明でわかりやすいことが重要。だから EncoderとPredictorで分けて実装するのでOK.

Agentで用いるときに、データはバッチ軸が常にないことが想定されるので、もし可能であればEncoderにバッチ軸がない場合の処理を追加してほしい。`InferenceWrapper`を自作する必要がありそうだからそこに書いてもいいけど。

実装してくれた上でこういうことをいうのは申し訳ないんだけど、コピペをするのはやめよう。時間がかかっても良いんだ。雑で巨大なコードをレビューするのは非常に大変で、抜け漏れを起こしやすい。僕は機械知能を大切にしたいから、焦って雑なコードを書くよりは遅れた方がいい。
ただ、タスク分散とかは計るから、**やばい時はなんとかやり切るよりも相談しよう**

File: ami/models/i_jepa.py [hunk 1]
```diff
@@ -0,0 +1,648 @@
...
+            trunc_normal_(m.weight, std=self.init_std)
+            if isinstance(m, nn.Linear) and m.bias is not None:
+                nn.init.constant_(m.bias, 0)
+        elif isinstance(m, nn.LayerNorm):
+            nn.init.constant_(m.bias, 0)
+            nn.init.constant_(m.weight, 1.0)
+        elif isinstance(m, nn.Conv2d):
+            trunc_normal_(m.weight, std=self.init_std)
+            if m.bias is not None:
+                nn.init.constant_(m.bias, 0)
```
[the remark this answers is not in the document]
Comment (zassou65535) [reply ami/models/i_jepa.py]: https://github.com/MLShukai/ami-vconf24/pull/35/commits/f21a1aaaa66cd70721885c831556cf595327a6c9 にて修正

File: ami/models/i_jepa.py [hunk 2]
```diff
@@ -0,0 +1,648 @@
+# Ref: https://github.com/facebookresearch/ijepa
+
+import math
+from typing import Optional
```
[the remark this answers is not in the document]
Comment (zassou65535) [reply ami/models/i_jepa.py]: https://github.com/MLShukai/ami-vconf24/pull/35/commits/d4b9c9c9bc162222282f2e72642acaabd6f9cf7a にて変更

File: ami/models/i_jepa.py [hunk 3]
```diff
@@ -0,0 +1,648 @@
...
+def repeat_interleave_batch(x: torch.Tensor, batch_size: int, repeat: int) -> torch.Tensor:
+    N = len(x) // batch_size
+    x = torch.cat(
+        [torch.cat([x[i * batch_size : (i + 1) * batch_size] for _ in range(repeat)], dim=0) for i in range(N)],
+        dim=0,
+    )
+    return x
+
+
+def apply_masks(x: torch.Tensor, masks: list[torch.Tensor]) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (zassou65535) [reply ami/models/i_jepa.py]: https://github.com/MLShukai/ami-vconf24/pull/35/commits/ef46cf6e1676bd43bb33bde8126363cb37e4639a
`select_mask_patches`に名称変更

File: ami/models/i_jepa.py [hunk 4]
```diff
@@ -0,0 +1,648 @@
...
+    """Drop paths (Stochastic Depth) per sample  (when applied in main path of
+    residual blocks)."""
+
+    def __init__(self, drop_prob: float):
+        super().__init__()
+        assert drop_prob > 0.0
+        self.drop_prob = drop_prob
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        return drop_path(x, self.drop_prob, self.training)
```
[the remark this answers is not in the document]
Comment (zassou65535) [reply ami/models/i_jepa.py]: https://github.com/MLShukai/ami-vconf24/pull/35/commits/dd41109ed1084e77d8545fe28e2bf0ea0fac57b4
で修正

File: ami/models/i_jepa.py [hunk 5]
```diff
@@ -0,0 +1,648 @@
...
+        self.apply(self._init_weights)
+        self.fix_init_weight()
+
+    def fix_init_weight(self) -> None:
+        def rescale(param: torch.Tensor, layer_id: int) -> None:
+            param.div_(math.sqrt(2.0 * layer_id))
+
+        for layer_id, layer in enumerate(self.predictor_blocks):
+            rescale(layer.attn.proj.weight.data, layer_id + 1)
+            rescale(layer.mlp.fc2.weight.data, layer_id + 1)
```
[the remark this answers is not in the document]
Comment (zassou65535) [reply ami/models/i_jepa.py]: https://github.com/MLShukai/ami-vconf24/pull/35/commits/b7c6042dc72e28c12d56b3f9133a1054b3c2bc0d
にて変更

**Comment by zassou65535** [comment]
suggestしてくれたcommitと、すぐに直せそうなところを修正した。
レビューしてくれた中でまだ修正されていないところもあるけど、詳細に解析したいところもあるから一回このプルリクはclose、分割してそっちで追って修正したいかも
